Application Version Compliance Report

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
Frontend
<img width="290" alt="image" src="https://github.com/user-attachments/assets/0973746e-17af-4f49-bb56-9fd719aa1ebf" />

Report
<img width="733" alt="image" src="https://github.com/user-attachments/assets/b14438fd-118d-4340-bb75-dfe01ddc5e92" />


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-16311

Review by commit. ( The commit history is basically how I improved the performance step by step )

This PR is to add a new Application Version Compliance Tile to the Enterprise Console.
This tile will iterate through all domain in this enterprise account, in each domain, iterate through all mobile workers, for each mobile worker, iterate through all their `last_build` in `last_builds`, (`last_builds` include the last build of the application they installed),  and see if the build is out of date, if it is, then append it to the report's row. 


## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->



### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Add App Version Compliance Report

jingcheng16 committed:
app_id can be None or empty string

jingcheng16 committed:
is_out_of_date expect string

jingcheng16 committed:
Exclude apps that has been deleted

jingcheng16 committed:
Exclude deleted app before wrap

jingcheng16 committed:
Extract necessary attributes directly instead of wrapping the raw document

This commit improved performance by 92%

jingcheng16 committed:
Use built_on when no last_released

jingcheng16 committed:
Rename wrapped_builds_cache to build_by_build_id

since we're no longer wrapping build doc

jingcheng16 committed:
Cache application name by id

This commit improved performance by 92%

jingcheng16 committed:
Don't query all domains object

jingcheng16 committed:
Refactor: reorder funcion

jingcheng16 committed:
Avoid wraping Application object

Improve performance by 20%

jingcheng16 committed:
Avoid wrap LastBuild

Improved performance by 30%

jingcheng16 committed:
Only fetch all builds if necessary

It increases num of calls to `get_app_builds` by 10%, but still saves a lot of time because most `get_app_builds` now only need to return first 10 builds

jingcheng16 committed:
Implement total calculation

jingcheng16 committed:
Refactor: DRY

jingcheng16 committed:
Rename variables and show up-to-date percentage instead of out-of-date percentage

nospame commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+            for build in last_builds:
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
```
Since this is iterating by domain, could a user have an app that's not in the passed in `apps`? A user here belongs to a domain, right, and should only have apps that belong to that same domain? I do see that this would also catch the situation where `app_id` is `None`.

jingcheng16 replied:
Iteration will skip when `last_builds` have build information for a deleted app.

nospame replied:
Forgot about deleted apps. Makes sense to me, then!

nospame commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+        Get the latest build version at the time
+
+        :param all_builds: List of raw build documents sorted by version in descending order
+        :param time: The date of the build version to compare against
```
Might be worth specifying what date format is expected here. Seems like it's whatever is returned by `DateTimeProperty.deserialize`.

jingcheng16 replied:
I thought date by default means a datetime object, rather than string?

nospame replied:
It's getting called with `build_version_date` that's comes from `DateTimeProperty.deserialize(build.get('build_version_date'))`. Is that a datetime object? I would prefer to see it specified in the docstring whether it is or not, so there's no ambiguity.

jingcheng16 replied:
Docstring updated with the refactoring: 25e5e521275eec772dda66abc24667216cda82c6

nospame commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
+                    all_builds = self.get_app_builds(domain, app_id)
+                    latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
```
It seems like `get_app_builds` is only used by `get_latest_build_version_at_time`. Rather than just retrying the query here, could that `if not` logic be moved (or removed) if `get_app_builds` is called directly by `get_latest_build_version_at_time`?

jingcheng16 replied:
Can you elaborate on this? Here what I'm trying to do is first get the latest 10 builds, and see if we can identify the latest version at use time from this 10 builds, if not, then query again to get all builds, and try to identify the latest version from all builds

nospame replied:
It makes sense what this is trying to do, I think what I'm getting at is that I think the retry logic should be handled in the method doing the fetching, and doesn't need to be seen by `_get_user_builds`. Maybe I'm overcomplicating it/trying to overabstract things. But I'm imagining something like:
```python
def get_latest_build_version_at_time(domain, app_id, time):
    builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
    latest_build = self._get_latest_build_version_if_present(builds, time)

    if latest_build is None:
        builds = self.get_app_builds(domain, app_id, start=self.INITIAL_QUERY_LIMIT)
        latest_build = self._get_latest_build_version_if_present(builds, time)

    return latest_build
```
Where `_get_latest_build_version_if_present` (or whatever naming) is handling the loop and returning `None` if the latest isn't in the builds that were passed in, and `get_app_builds` is modified to take a `start` parameter or something similar, so it doesn't return the same initial 10 builds along with the rest of them. Then `get_latest_build_version_at_time` will always return the actual latest build version.

gherceg replied:
+1 to separating out that logic. A helper method should be capable of doing everything to get the latest build version.

jingcheng16 replied:
Refactored in 25e5e521275eec772dda66abc24667216cda82c6, also updated the docstring!

nospame commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+        return None
+
+
+def _format_percentage_for_enterprise_tile(dividend, divisor):
```
Is it correct to have this marked as "private" with the leading `_`? It seems like it's used in a few different places in this file now.

jingcheng16 replied:
My understanding of private function is it will only be used in the file, never be imported. So use it multiple times in one file seems fine?

nospame replied:
Okay, I read up a little on this and I think you're correct!

nospame reviewed (commented):
I think this looks good! Did leave some context/structural/reasoning questions.

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+                _('Version in Use'), _('Last Used [UTC]')]
+
+    @property
+    def rows(self):
```
Yes, the parent `EnterpriseReport` class's `rows` is iterate through all the domain object, to get all the domain object, [it iterate through all domain's name and get the domain object.](https://github.com/dimagi/commcare-hq/blob/848cf09907dadc7919d3727a602d50fb4c5c5b50/corehq/apps/enterprise/enterprise.py#L133-L134), it takes time for large enterprise account have more than 400 domains. Since this tile, only domain name is required, I prefer don't call `domains`.

nospame replied:
Got it, I missed that. Looks good then.

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+        total_out_of_date_rows = 0
+        total_last_builds = 0
+
+        for domain in self.account.get_domains():
```
This is done intentionally to avoid get domain objects for each domain

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+                'reporting_metadata.last_builds',
+            ]))
+        for user in user_query.run().hits:
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
```
I'm not sure. I think I'm just doing it in case it is missing. I saw we're doing it [here](https://github.com/dimagi/commcare-hq/blob/848cf09907dadc7919d3727a602d50fb4c5c5b50/corehq/apps/reports/standard/deployments.py#L336) for `reporting_metadata` as well. There is no harm to add a fallback value.

nospame replied:
I'm not sure I agree in principle. To me, adding a fallback value indicates that _there is definitely_ a situation where the key can be missing, versus considering it an error state where something has gone wrong. I'd call this a non-blocking comment, though.

jingcheng16 committed:
text update and reformat

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +639,87 @@ def rows_for_domain(self, domain_obj):
+
+class EnterpriseAppVersionComplianceReport(EnterpriseReport):
+    title = gettext_lazy('Application Version Compliance')
+    total_description = gettext_lazy('% of Applications Up-to-Date Across All Mobile Workers')
```
Thank you!! b024957ecb43d0261689e096d0e669e7b6566c21

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+    def headers(self):
+        return [_('Mobile Worker'), _('Project Space'), _('Application'),
+                _('Latest Version Available at Submission'),
+                _('Version in Use'), _('Last Used [UTC]')]
```
Didn't commit suggestion so I can combine this commit with the text update: b024957ecb43d0261689e096d0e669e7b6566c21

jingcheng16 committed:
Rename column

Because this tile based on `LastBuild`, build will be updated by either a form submission, or a sync, or mobile heartbeat.

jingcheng16 committed:
Add ApplicationVersionCompliance api

gherceg commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,153 @@ def rows_for_domain(self, domain_obj):
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
```
It seems unfortunate that we have to fetch all builds to find the latest version available at the time of use. Is it possible to have a more targeted query using the built_on and/or last_released properties? The downside is that would probably be less effective for the current caching mechanism and in turn might result in more ES queries which could be worse, even if they are targeted. Is that something you have explored?

jingcheng16 replied:
A more targeted query takes 200 sec in total...I don't understand why it takes that long...
```
app_es = (
    AppES()
    .domain(domain)
    .is_build()
    .app_id(app_id)
    .sort('version', desc=True)
    .is_released()
    .source(['_id', 'version'])
    .filter(
        OR(
            range_filter('last_released', lte=before_date),
            range_filter('built_on', lte=before_date)
        )
    )
    .size(1)
)
```

gherceg replied:
Jing and I offlined about this and I believe cleared this one up. Thanks for doing the exploration on this and sorry that it didn't help anything 😅

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -639,6 +639,7 @@ def rows_for_domain(self, domain_obj):
 class EnterpriseAppVersionComplianceReport(EnterpriseReport):
     title = gettext_lazy('Application Version Compliance')
     total_description = gettext_lazy('% of Applications Up-to-Date Across All Mobile Workers')
+    INITIAL_QUERY_LIMIT = 10
```
I think normally the version should be not too out-of-date, so I thought 10 might be a good number...

jingcheng16 committed:
Refactor: Move get_app_builds inside get_latest_build_version_at_time

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+                if not latest_version:
+                    all_builds = self.get_app_builds(domain, app_id)
+                    latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                yield user, build, latest_version
```
_get_user_builds is also used in https://github.com/dimagi/commcare-hq/blob/25e5e521275eec772dda66abc24667216cda82c6/corehq/apps/enterprise/enterprise.py#L686-L698. I need those info for each row. Maybe I can return username instead of user?

jingcheng16 committed:
Refactor: rename app_id to app_ids

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,153 @@ def rows_for_domain(self, domain_obj):
+        total_last_builds = 0
+        total_out_of_date = 0
+
+        for user, build, latest_version in self._get_user_builds(domain, apps):
```
c446ad1ebfcd3440147dc3172f9a330afa07936d

mjriley reviewed (commented):
Looks good. Mostly some minor suggestions that we could do to clarify some of the code, but do have some questions over whether some optimizations are necesssary

jingcheng16 committed:
Clear the cache for initial 10 builds

jingcheng16 committed:
Refactor: renaming variable and text update

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+
+        if latest_build is None:
+            builds = self.get_app_builds(domain, app_id, start=self.INITIAL_QUERY_LIMIT)
+            latest_build = self._find_latest_build_version_from_builds(builds, time)
```
I did this experiment:
<img width="672" alt="image" src="https://github.com/user-attachments/assets/bdb20479-8a9e-4ffd-aff9-7aa20426f38c" />
I plan to change Initial Query Limit to 15. 
Where should I document it?

jingcheng16 replied:
But since we are not displaying a metric in front end. I guess I can remove that logic too. 08443df3c5703913c93a4a71b0f007bc9f919881

jingcheng16 committed:
Remove Initial Query Limit since it makes the logic complicated

It will worsen the performance by 20%

jingcheng16 committed:
Return a dictionary in _get_user_builds

biyeun commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+
+    def total_for_domain(self, domain):
+        app_ids = get_app_ids_in_domain(domain)
+        total_last_builds = 0
```
looks like this was updated in a different location

jingcheng16 replied:
This is removed in 12a4ff624b06a8581092cf70b8c11f085340e6b4 since we're not displaying metric

biyeun commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+                yield user, build, latest_version
+
+    def get_latest_build_version_at_time(self, domain, app_id, time):
+        """
+        Get the latest build version available at the given time.
+
+        :param domain: The domain of the app
+        :param app_id: The application id
+        :param time: A datetime object representing the date of the build version to compare against
+        :return: The latest build version available at the given date
+        """
```
from review: rename to `get_latest_build_version`, and rename last parameter to `at_datetime` and remove docstring? unless folks have other ideas.

jingcheng16 replied:
5bb7b5a94a024e004b718c4399b9da7892252b93

biyeun commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+
+        return self.builds_by_app_id[app_id]
+
+    def _find_latest_build_version_from_builds(self, all_builds, time):
+        """
+        Get the latest build version at the time
+
+        :param all_builds: List of raw build documents sorted by version in descending order
+        :param time: A datetime object representing the date of the build version to compare against
+        :return: The latest build version available at the given date.
+        """
```
from review: rename `time` to `at_datetime`, remove doctstring. add tests

jingcheng16 committed:
Always show '--' when display metric

jingcheng16 committed:
Refactor: renaming parameter so the function is self-explanatory and docstring can be removed

jingcheng16 committed:
Refactor: rename variable to make it more self-explanatory

jingcheng16 committed:
Refactor: extract the logic in _find_latest_build_version_from_builds to a function

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+                    'version': build_doc['version'],
+                    'last_released': DateTimeProperty.deserialize(released_date)
+                }
+                self.build_by_build_id[build_id] = build_info
```
8a4accd4b46b7b5cd4469ad4ea4eaea26bfa9bd0

jingcheng16 committed:
Refactor: check the cache first and return immediately

jingcheng16 commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+                app_es = app_es.size(limit)
+            self.builds_by_app_id[app_id] = app_es.run().hits
+
+        return self.builds_by_app_id[app_id]
```
c79f198120fb8001c67bc05d5274c5f2a35deace

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -666,51 +665,31 @@ def rows(self):
-        total_up_to_date = total_last_builds - total_out_of_date_rows
-
-        return _format_percentage_for_enterprise_tile(total_up_to_date, total_last_builds)
+        return '--'
```
Can we add a comment explaining why we're skipping this display for now?

jingcheng16 replied:
3a0985a9bbb38580f6bf7dd01267f0902ddd6478

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -727,73 +706,57 @@ def _get_user_builds(self, domain, app_ids):
 
+    def _get_build_info(self, build_doc):
+        build_id = build_doc['_id']
+        build_info = self.build_info_cache.get(build_id)
+        if not build_info:
+            # last_released is added in 2019, build before 2019 don't have this field
+            # TODO: have a migration to populate last_released from built_on
+            # Then this code can be modified to use last_released only
+            released_date = build_doc.get('last_released') or build_doc['built_on']
+            build_info = {
+                'version': build_doc['version'],
+                'last_released': DateTimeProperty.deserialize(released_date)
+            }
+            self.build_info_cache[build_id] = build_info
```
Can you explain the role of `build_info_cache` here? It looks like we're already caching the raw results from the elasticsearch query. Then, within `_find_latest_build_version_from_builds`, we just want to turn that raw doc into  something we can use to determine whether or not this build's release date was prior to the datetime specified. A conversion function doesn't sound like it would need a cache. Was this something that came up while profiling? If we are caching these results to avoid repeated calls to `DateTimeProperty.deserialize`, perhaps that is something we can do when we first cache the elasticsearch results?

jingcheng16 replied:
`build_info_cache` is to cache all required information of a build, mainly version number and last released date. I did this because I realized deserialize also have an impact on performance, so I want to do less deserialization.
> perhaps that is something we can do when we first cache the elasticsearch results?

This will result in deserialize unnecessary build, for example, if an app already have 1000 builds, then it is very likely the first 500 build's last released date will never be required in the comparison then no need to deserialize.

mjriley replied:
If you haven't seen it yet, there's an infamous statement about ["Premature optimization is the root of all evil"](https://wiki.c2.com/?PrematureOptimization). 

I wrote up a quick performance test to gauge how big of an impact deserialization had:
```
from datetime import datetime
from dimagi.utils.dates import safe_strftime
from dimagi.utils.parsing import ISO_DATETIME_FORMAT
from dimagi.ext.couchdbkit import DateTimeProperty
import random

TIMESTAMP_FLOOR = 1_000_000_000
TIMESTAMP_CEIL = 1_730_000_000

def create_datetime_string():
    return safe_strftime(
        datetime.fromtimestamp(
            random.randrange(TIMESTAMP_FLOOR, TIMESTAMP_CEIL)
        ), ISO_DATETIME_FORMAT)

def create_n_dates(n):
    return [create_datetime_string() for i in range(n)]

def time_deserialize():
    num_elements = 1000
    all_dates = create_n_dates(num_elements)
    deserialized_dates = [None] * num_elements

    start = datetime.now()
    for i in range(num_elements):
        deserialized_dates[i] = DateTimeProperty.deserialize(all_dates[i])
    end = datetime.now()

    print(f'Took: {end - start}')

```

TL;DR -- the time to deserialize 1K datestrings, on my machine, took 10K to 15K microseconds -- i.e. milliseconds.

From the current code, it appears `build_info_cache` only cares about the build number and last released date -- I don't see any other data being used. Given that deserialization has virtually no performance impact, I think it makes the code cleaner for `get_app_builds` to return only the needed information (version, last_released_date) and do the deserialization immediately. This allows us to get rid of `build_info_cache`

mjriley replied:
If you haven't seen it yet, there's an infamous statement about ["Premature optimization is the root of all evil"](https://wiki.c2.com/?PrematureOptimization). 

I wrote up a quick performance test to gauge how big of an impact deserialization had:
```
from datetime import datetime
from dimagi.utils.dates import safe_strftime
from dimagi.utils.parsing import ISO_DATETIME_FORMAT
from dimagi.ext.couchdbkit import DateTimeProperty
import random

TIMESTAMP_FLOOR = 1_000_000_000
TIMESTAMP_CEIL = 1_730_000_000

def create_datetime_string():
    return safe_strftime(
        datetime.fromtimestamp(
            random.randrange(TIMESTAMP_FLOOR, TIMESTAMP_CEIL)
        ), ISO_DATETIME_FORMAT)

def create_n_dates(n):
    return [create_datetime_string() for i in range(n)]

def time_deserialize():
    num_elements = 1000
    all_dates = create_n_dates(num_elements)
    deserialized_dates = [None] * num_elements

    start = datetime.now()
    for i in range(num_elements):
        deserialized_dates[i] = DateTimeProperty.deserialize(all_dates[i])
    end = datetime.now()

    print(f'Took: {end - start}')

```

TL;DR -- the time to deserialize 1K datestrings, on my machine, took 10K to 15K microseconds -- i.e. milliseconds.

From the current code, it appears `build_info_cache` only cares about the build number and last released date -- I don't see any other data being used. Given that deserialization has virtually no performance impact, I think it makes the code cleaner for `get_app_builds` to return

jingcheng16 replied:
I finally find the commit where I decide to deserialize separately without wrapping the raw document: https://github.com/dimagi/commcare-hq/commit/f94b4b31014a3b7da4c709e1361746aa534fc7e3
I think because I cached the wrapped result initially, so when I change to only deserialize the datetime without wrapping it, I never thought of getting rid of the cache. 
I will put up a PR to remove the cache.

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
+            if is_out_of_date(str(build['build_version']), str(latest_version)):
+                app_id = build['app_id']
+                if app_id not in app_name_by_id:
+                    app_name_by_id[app_id] = Application.get_db().get(app_id).get('name')
```
This definitely seems like the performance vs readability discussion. Now that it appears we no longer have the strict performance concerns (due to not populating the summary data), do you think we should emphasize the readability/maintainability side rather than the 20% performance gain?

jingcheng16 replied:
Can you explain more on why `Application.get_db()` have readability issue?

mjriley replied:
This is the Law of Demeter, that I linked above. When modules start knowing about more than 1 level of internal structure of another module, things tend to get brittle -- unexpected errors crop up and the code is more difficult to maintain. I'm sure the link I provided above has a more full description. `Application.get_db().get(app_id).get('name')` means that this enterprise module essentially knows about the data access layer. Not just how to query the database (`Application.get_db().get(app_id)`) but also about how the database internally represents an Application. The application layer only generally wants to deal with fully constructed `Application` objects, which gives us flexibility -- `name` could be stored in an arbitrary manner in the database, but anyone interacting with the `Application` object wouldn't need to know those details. With this code, we do.

mjriley commented on corehq/apps/enterprise/enterprise.py:
```diff
@@ -666,51 +665,31 @@ def rows(self):
-        for user, build, latest_version in self._get_user_builds(domain, app_ids):
-            if is_out_of_date(str(build['build_version']), str(latest_version)):
-                app_id = build['app_id']
+        for row_data in self._get_user_builds(domain, app_ids):
+            version_in_use = str(row_data['build']['build_version'])
+            latest_version = str(row_data['latest_version'])
+            if is_out_of_date(version_in_use, latest_version):
+                app_id = row_data['build']['app_id']
```
I don't know if this is something you need to act on, but wanted to bring it up -- I think I struggled to figure out what was going on here, because `_get_user_builds` doesn't sound like it's going to be returning latest version information, and because `row_data` doesn't really suggest what is being dealt with.

In pseudo-code, I'd imagine this is what we're trying to do:
```
retrieve app-last-usage information for all users in a domain
for each app-last-usage:
  retrieve the most current version at the time the app was last used
  if that version is not the same version as is in our information, then
    add a row to our report with the relevant details
```

I'm not sure the above is any better, but if you feel there's some confusion with what is going on with `row_data`, perhaps its worth addressing

jingcheng16 replied:
I think I rename it to `row_data` because `build['build']['build_version]` looks weird. What do you think of change line 675 to `for build_and_latest_version in self.all_last_builds_with_latest_version(domain, app_ids):
`

mjriley replied:
That seems like a good improvement

jingcheng16 replied:
Thanks! df144d43d8404bb056f1f5370cf4176ca312f11a

mjriley reviewed (commented):
I think I have some remaining questions on our performance tradeoffs and some of the caching

jingcheng16 committed:
Merge branch 'master' into jc/app_version_compliance_report

jingcheng16 committed:
Add comment

jingcheng16 committed:
Merge branch 'master' into jc/app_version_compliance_report

jingcheng16 committed:
Refactor: rename to improve readability

jingcheng16 committed:
Update the descriptor

mjriley reviewed (approved):
I'm approving to allow this tile to be live on production. I'm convinced this code works, but I hope we can clean up this code in the future to make it more readable and maintainable.
